### PR TITLE
HIVE-25842: Reimplement delta file metric collection

### PR DESCRIPTION
**Pull request by lcspinter**
### What changes were proposed in this pull request?
Move delta metric collection from Tez side to compaction side. All delta file metrics are collected during initiator, worker and cleaner phase.

### Why are the changes needed?
Metrics are collected only when a Tez query runs a table (select * and select count( * ) don't update the metrics)
Metrics aren't updated after compaction or cleaning after compaction, so users will probably see "issues" with compaction (like many active or obsolete or small deltas) that don't exist.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Manual test, unit test

### CODE REVIEW

File: ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/metrics/DeltaFilesMetricReporter.java [hunk 1]
```diff
@@ -398,13 +156,6 @@ private static long getBaseSize(AcidDirectory dir) throws IOException {
...
 
-  private static long getModificationTime(AcidUtils.ParsedDirectory dir, FileSystem fs) throws IOException {
-    return dir.getFiles(fs, Ref.from(false)).stream()
-      .map(HadoopShims.HdfsFileStatusWithId::getFileStatus)
-      .mapToLong(FileStatus::getModificationTime)
-      .max()
-      .orElse(new Date().getTime());
-  }
 
   private static long getDirSize(AcidUtils.ParsedDirectory dir, FileSystem fs) throws IOException {
```
Comment (klcopp) [inline ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/metrics/DeltaFilesMetricReporter.java:160]: I'm a bit concerned about this method. Do you think it might slow down the Initiator/Worker/Cleaner?

File: ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/metrics/DeltaFilesMetricReporter.java [hunk 2]
```diff
@@ -512,7 +284,177 @@ private void shutdown() {
...
+          numNewDeltas++;
+        }
+
+        // recalculate the delta count
+        CompactionMetricsDataStruct prevDelta =
+            client.getCompactionMetricsData(dbName, tableName, partitionName, CompactionMetricsMetricType.NUM_DELTAS)
+                .getData();
+        int deltaNum = numNewDeltas;
+        if (prevDelta != null) {
+          deltaNum += prevDelta.getMetricvalue() - directory.getCurrentDirectories().size();
```
Comment (klcopp) [inline ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/metrics/DeltaFilesMetricReporter.java:379]: There is a high chance that directory.getCurrentDirectories().size() > prevDelta.getMetricvalue() , i.e. more deltas were inserted since the initiator queued this compaction... in this case deltaNum shouldn't be changed

File: ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/Cleaner.java [hunk 3]
```diff
@@ -396,7 +423,9 @@ private boolean removeFiles(String location, ValidWriteIdList writeIdList, Compa
     }
     StringBuilder extraDebugInfo = new StringBuilder("[").append(obsoleteDirs.stream()
         .map(Path::getName).collect(Collectors.joining(",")));
-    return remove(location, ci, obsoleteDirs, true, fs, extraDebugInfo);
+    boolean success = remove(location, ci, obsoleteDirs, true, fs, extraDebugInfo);
+    updateDeltaFilesMetrics(ci.dbname, ci.tableName, ci.partName, dir.getObsolete().size());
```
Comment (klcopp) [inline ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/Cleaner.java:427]: should probably be only updated if success == true (I believe this means that at least 1 dir was removed)

Comment (klcopp) [reply ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/Cleaner.java:427]: More readable solution : Update only if dir.getObsolete().size() > 0

File: ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/Cleaner.java [hunk 4]
```diff
@@ -396,7 +423,9 @@ private boolean removeFiles(String location, ValidWriteIdList writeIdList, Compa
...
+    return success;
```
Comment (klcopp) [inline ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/Cleaner.java:428]: - Base directories and original files may be in the list of obsolete files the Cleaner deletes. Do you think it would be worth filtering the obsolete/aborted files’ names for “delta”?

File: ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/Cleaner.java [hunk 5]
```diff
@@ -87,14 +106,15 @@ public void init(AtomicBoolean stop) throws Exception {
     cleanerExecutor = CompactorUtil.createExecutorWithThreadFactory(
             conf.getIntVar(HiveConf.ConfVars.HIVE_COMPACTOR_CLEANER_THREADS_NUM),
             COMPACTOR_CLEANER_THREAD_NAME_FORMAT);
+    metricsEnabled = MetastoreConf.getBoolVar(conf, MetastoreConf.ConfVars.METRICS_ENABLED) &&
```
Comment (deniskuzZ) [inline ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/Cleaner.java:109]: with that change you won't be able to enable/disable metrics collection at runtime, is it intentional?

File: ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/Cleaner.java (hunk 3 above)
Comment (deniskuzZ) [inline ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/Cleaner.java:427]: there are a few remove methods, see line #338 (soft-drop partition), updateDeltaFilesMetrics should be called there as well, or if possible we should move it inside of remove at line #431

File: ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/Initiator.java [hunk 6]
```diff
@@ -275,6 +276,13 @@ public void init(AtomicBoolean stop) throws Exception {
     compactionExecutor = CompactorUtil.createExecutorWithThreadFactory(
             conf.getIntVar(HiveConf.ConfVars.HIVE_COMPACTOR_REQUEST_QUEUE),
             COMPACTOR_INTIATOR_THREAD_NAME_FORMAT);
+    metricsEnabled = MetastoreConf.getBoolVar(conf, MetastoreConf.ConfVars.METRICS_ENABLED) &&
```
Comment (deniskuzZ) [inline ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/Initiator.java:279]: same as above, q is whether we need to support runtime changes to the metrics configs

File: ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/Initiator.java [hunk 7]
```diff
@@ -331,6 +339,12 @@ private boolean foundCurrentOrFailedCompactions(ShowCompactResponse compactions,
     }
     return false;
   }
+  
+  private void updateDeltaFilesMetrics(AcidDirectory directory, String dbName, String tableName, String partName) {
```
Comment (deniskuzZ) [inline ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/Initiator.java:343]: i think, this method could be extracted to a common place to be used by Cleaner and Initiator, like MetaStoreCompactorThread

File: ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/Initiator.java [hunk 8]
```diff
@@ -381,17 +397,19 @@ public CompactionType run() throws Exception {
     }
   }
 
-  private CompactionType determineCompactionType(CompactionInfo ci, ValidWriteIdList writeIds,
-                                                 StorageDescriptor sd, Map<String, String> tblproperties)
+  private AcidDirectory getAcidDirectory(StorageDescriptor sd,ValidWriteIdList writeIds) throws IOException {
```
Comment (deniskuzZ) [inline ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/Initiator.java:400]: space

File: ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/Worker.java [hunk 9]
```diff
@@ -671,6 +679,13 @@ private String getWorkerId() {
     return name.toString();
   }
 
+  private void updateDeltaFilesMetrics(AcidDirectory directory, String dbName, String tableName, String partName,
```
Comment (deniskuzZ) [inline ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/Worker.java:682]: move to the common place if possible. updateDeltaFilesMetrics could be triggered by multiple threads leading to resource starvation, think about introducing connection pool

File: ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/metrics/DeltaFilesMetricReporter.java (hunk 1 above)
Comment (lcspinter) [reply ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/metrics/DeltaFilesMetricReporter.java:161]: It will slow it down, but there is no other way we could calculate the directory size.

File: ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/metrics/DeltaFilesMetricReporter.java (hunk 2 above)
Comment (lcspinter) [reply ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/metrics/DeltaFilesMetricReporter.java:385]: That is true, but in that case shouldn't we clean up the active delta metrics, if that falls behind the threshold?

File: ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/metrics/DeltaFilesMetricReporter.java [hunk 10]
```diff
@@ -512,7 +284,177 @@ private void shutdown() {
...
+      int numSmallDeltas = 0;
+
+      for (AcidUtils.ParsedDelta delta : dir.getCurrentDirectories()) {
+        long deltaSize = getDirSize(delta, dir.getFs());
+        if (baseSize != 0 && deltaSize / (float) baseSize < deltaPctThreshold) {
+          numSmallDeltas++;
+        }
+      }
+
+      int numObsoleteDeltas = dir.getObsolete().size();
```
[the remark this answers is not in the document]
Comment (lcspinter) [reply ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/metrics/DeltaFilesMetricReporter.java:310]: Do we want to calculate aborted directories in the obsolete delta count?

File: ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/Worker.java [hunk 11]
```diff
@@ -142,6 +144,9 @@ public void init(AtomicBoolean stop) throws Exception {
     super.init(stop);
     this.workerName = getWorkerId();
     setName(workerName);
+    metricsEnabled = MetastoreConf.getBoolVar(conf, MetastoreConf.ConfVars.METRICS_ENABLED) &&
+        MetastoreConf.getBoolVar(conf, MetastoreConf.ConfVars.METASTORE_ACIDMETRICS_EXT_ON) &&
+        MetastoreConf.getBoolVar(conf, MetastoreConf.ConfVars.COMPACTOR_INITIATOR_ON);
```
[the remark this answers is not in the document]
Comment (klcopp) [reply ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/Worker.java:152]: There is a possibility that COMPACTOR_INITIATOR_ON==false on a given HS2 instance (even if the Initiator / Cleaner are running in some HMS somewhere else).

There's already a risk that MetastoreConf.ConfVars.METRICS_ENABLED == false on any given HS2, which means that only the Initiator and Cleaner are updating the metrics, which means the metric values are incorrect. Adding another config (COMPACTOR_INITIATOR_ON) just increases this risk.

File: ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/metrics/DeltaFilesMetricReporter.java (hunk 1 above)
Comment (klcopp) [reply ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/metrics/DeltaFilesMetricReporter.java:161]: Do you think that collecting metrics about small deltas is worth the amount this might slow down the Initiator/Worker/Cleaner?

File: ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/Cleaner.java (hunk 4 above)
Comment (klcopp) [reply ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/Cleaner.java:429]: Let me rephrase then (discussion of aborted directories below):
Base directories and original files may be in the list of obsolete files the Cleaner deletes. Do you think it would be worth filtering the obsolete files’ names for “delta”?

File: ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/metrics/DeltaFilesMetricReporter.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (klcopp) [reply ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/metrics/DeltaFilesMetricReporter.java:310]: I'm seriously wondering if we should.
Cons:
- the metric name is "obsolete deltas" would == obsolete + aborted deltas
- We already have metrics about the amount of aborts in the system (gotten from metadata)
Pros:
- The Cleaner's job is to remove aborted directories as well ; so including metrics about aborted directories would help with observability of Cleaner health
- Aborted directories can clog up the file system just as much as obsolete directories

File: standalone-metastore/metastore-server/src/main/sql/derby/hive-schema-4.0.0.derby.sql [hunk 12]
```diff
@@ -661,6 +661,16 @@ CREATE TABLE COMPLETED_COMPACTIONS (
 
 CREATE INDEX COMPLETED_COMPACTIONS_RES ON COMPLETED_COMPACTIONS (CC_DATABASE,CC_TABLE,CC_PARTITION);
 
+-- HIVE-25842
+CREATE TABLE COMPACTION_METRICS_CACHE (
+  CMC_DATABASE varchar(128) NOT NULL,
+  CMC_TABLE varchar(128) NOT NULL,
+  CMC_PARTITION varchar(767),
+  CMC_METRIC_TYPE varchar(128) NOT NULL,
+  CMC_METRIC_VALUE integer NOT NULL,
```
[the remark this answers is not in the document]
Comment (klcopp) [reply standalone-metastore/metastore-server/src/main/sql/derby/hive-schema-4.0.0.derby.sql:670]: Currently yes, but if the use of this table is expanded in the future, do you think there's any chance we'll want to allow nulls?

File: ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/metrics/DeltaFilesMetricReporter.java [hunk 13]
```diff
@@ -435,55 +187,71 @@ private void initObjectsForMetrics() throws Exception {
...
+              Collectors.toMap(item -> getDeltaCountKey(item.getDbName(), item.getTblName(), item.getPartitionName()),
+                  CompactionMetricsData::getMetricValue));
+          smallDeltaObject.updateAll(smallDeltasMap);
+
+          Map<String, Integer> obsoleteDeltasMap = deltas.stream()
+              .filter(d -> d.getMetricType() == CompactionMetricsData.MetricType.NUM_OBSOLETE_DELTAS).collect(
+              Collectors.toMap(item -> getDeltaCountKey(item.getDbName(), item.getTblName(), item.getPartitionName()),
+                  CompactionMetricsData::getMetricValue));
+          obsoleteDeltaObject.updateAll(obsoleteDeltasMap);
+        } catch (MetaException e) {
```
Comment (klcopp) [inline ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/metrics/DeltaFilesMetricReporter.java:221]: Maybe catch all Throwables here just in case? (and also in run())

File: ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/Worker.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (lcspinter) [reply ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/Worker.java:152]: I see your point now. I will remove this additional config check.

File: standalone-metastore/metastore-server/src/main/sql/derby/hive-schema-4.0.0.derby.sql (hunk 12 above)
[the remark this answers is not in the document]
Comment (lcspinter) [reply standalone-metastore/metastore-server/src/main/sql/derby/hive-schema-4.0.0.derby.sql:670]: If for some reason, later on, we want to allow NULL values in this column (I can't think of any reason right now why would we do that), the NOT NULL -> NULL conversion is backward compatible change.

File: ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/metrics/DeltaFilesMetricReporter.java (hunk 1 above)
Comment (lcspinter) [reply ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/metrics/DeltaFilesMetricReporter.java:161]: I think we should collect the small delta metrics because they can indicate issues with streaming (I wish I had such a metric when I investigated the Northfolk escalation :) )
Also, we are doing the same calculation in the `Initiator` code (`Initiator#sumDirSize`) to determine the compaction type.

File: ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/metrics/DeltaFilesMetricReporter.java (hunk 13 above)
Comment (lcspinter) [reply ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/metrics/DeltaFilesMetricReporter.java:221]: Right, I will do that.

File: ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/metrics/DeltaFilesMetricReporter.java (hunk 2 above)
Comment (klcopp) [reply ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/metrics/DeltaFilesMetricReporter.java:385]: At line 378, deltaNum== numNewDeltas=={0,1,2} If prevDelta.getMetricvalue() - directory.getCurrentDirectories().size() < 0 (because new deltas were inserted since the Initiator updated the metric) then deltaNum should still be =={0,1,2} at line 380.
We'd still clean up the active delta metrics, if the threshold > {0,1,2}.

Comment (lcspinter) [reply ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/metrics/DeltaFilesMetricReporter.java]: I corrected this part.

File: ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/metrics/DeltaFilesMetricReporter.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (lcspinter) [reply ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/metrics/DeltaFilesMetricReporter.java]: Included aborted directories and filtered out original and base directories from the obsolete delta count.

File: ql/src/test/org/apache/hadoop/hive/ql/txn/compactor/TestDeltaFilesMetrics.java [hunk 14]
```diff
@@ -39,199 +45,396 @@
...
+    LockResponse res = txnHandler.lock(req);
+    Assert.assertEquals(LockState.ACQUIRED, res.getState());
+
+    long writeid = allocateWriteId(dbName, tblName, txnid);
+    Assert.assertEquals(24, writeid);
+    txnHandler.commitTxn(new CommitTxnRequest(txnid));
+
+    startInitiator();
+
+    TimeUnit.SECONDS.sleep(1);
```
[the remark this answers is not in the document]
Comment (lcspinter) [reply ql/src/test/org/apache/hadoop/hive/ql/txn/compactor/TestDeltaFilesMetrics.java:133]: Done.

File: ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/metrics/DeltaFilesMetricReporter.java [hunk 15]
```diff
@@ -310,81 +143,6 @@ private static String getDeltaCountKey(String dbName, String tableName, String p
...
-        Arrays.stream(DeltaFilesMetricType.values()).filter(type -> conf.get(type.name()) != null)
-            .forEach(type -> jobConf.set(type.name(), conf.get(type.name())));
-      } catch (Exception e) {
-        LOG.warn("Couldn't back propagate Delta metrics to jobConf object", e);
-      }
-    }
-  }
 
   private static long getBaseSize(AcidDirectory dir) throws IOException {
     long baseSize = 0;
```
[the remark this answers is not in the document]
Comment (lcspinter) [reply ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/metrics/DeltaFilesMetricReporter.java]: removed

File: ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/metrics/DeltaFilesMetricReporter.java [hunk 16]
```diff
@@ -512,7 +284,177 @@ private void shutdown() {
...
+      for (AcidUtils.ParsedDelta delta : dir.getCurrentDirectories()) {
+        long deltaSize = getDirSize(delta, dir.getFs());
+        if (baseSize != 0 && deltaSize / (float) baseSize < deltaPctThreshold) {
+          numSmallDeltas++;
+        }
+      }
+
+      int numObsoleteDeltas = dir.getObsolete().size();
+
+      if (numDeltas > deltasThreshold) {
```
[the remark this answers is not in the document]
Comment (lcspinter) [reply ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/metrics/DeltaFilesMetricReporter.java:288]: fixed

File: ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/metrics/DeltaFilesMetricReporter.java [hunk 17]
```diff
@@ -139,157 +92,37 @@ public static DeltaFilesMetricReporter getInstance() {
...
     initObjectsForMetrics();
 
     ThreadFactory threadFactory =
         new ThreadFactoryBuilder().setDaemon(true).setNameFormat("DeltaFilesMetricReporter %d").build();
-    executorService = Executors.newSingleThreadScheduledExecutor(threadFactory);
-    executorService.scheduleAtFixedRate(new ReportingTask(), 0, reportingInterval, TimeUnit.SECONDS);
+    reporterExecutorService = Executors.newSingleThreadScheduledExecutor(threadFactory);
+    reporterExecutorService.scheduleAtFixedRate(new ReportingTask(txnHandler), 0, reportingInterval, TimeUnit.SECONDS);
 
     LOG.info("Started DeltaFilesMetricReporter thread");
```
[the remark this answers is not in the document]
Comment (lcspinter) [reply ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/metrics/DeltaFilesMetricReporter.java:112]: Could you please point me to the original one? I cannot find it.

File: ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/metrics/DeltaFilesMetricReporter.java [hunk 18]
```diff
@@ -34,99 +26,60 @@
...
-import static org.apache.hadoop.hive.ql.txn.compactor.metrics.DeltaFilesMetricReporter.DeltaFilesMetricType.NUM_SMALL_DELTAS;
-
 /**
  * Collects and publishes ACID compaction related metrics.
- * Everything should be behind 2 feature flags: {@link HiveConf.ConfVars#HIVE_SERVER2_METRICS_ENABLED} and
+ * Everything should be behind 2 feature flags: {@link MetastoreConf.ConfVars#METRICS_ENABLED} and
  * {@link MetastoreConf.ConfVars#METASTORE_ACIDMETRICS_EXT_ON}.
- * First we store the information in the jobConf, then in Tez Counters, then in a cache stored here, then in a custom
- * MBean.
+ * First we store the information in the HMS backend DB COMPACTION_METRICS_CACHE table, then in a custom MBean.
```
[the remark this answers is not in the document]
Comment (lcspinter) [reply ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/metrics/DeltaFilesMetricReporter.java:64]: Done

File: ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/metrics/DeltaFilesMetricReporter.java [hunk 19]
```diff
@@ -512,7 +260,168 @@ private void shutdown() {
...
+        MetastoreConf.ConfVars.METASTORE_DELTAMETRICS_OBSOLETE_DELTA_NUM_THRESHOLD);
+    try {
+      // we have an instance of the AcidDirectory before the compaction worker was started
+      // from this we can get how many delta directories existed
+      // the previously active delta directories are now moved to obsolete
+      int numObsoleteDeltas = directory.getCurrentDirectories().size();
+      if (numObsoleteDeltas > obsoleteDeltasThreshold) {
+        updateMetrics(dbName, tableName, partitionName, CompactionMetricsMetricType.NUM_OBSOLETE_DELTAS,
+            numObsoleteDeltas, client);
+      }
```
[the remark this answers is not in the document]
Comment (lcspinter) [reply ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/metrics/DeltaFilesMetricReporter.java:326]: Corrected.

File: ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/metrics/DeltaFilesMetricReporter.java (hunk 17 above)
[the remark this answers is not in the document]
Comment (klcopp) [reply ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/metrics/DeltaFilesMetricReporter.java:116]: Never mind, I had reading problems :D

**Review by klcopp — APPROVED** [review]

File: ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/Cleaner.java [hunk 20]
```diff
@@ -412,7 +421,11 @@ private boolean removeFiles(String location, ValidWriteIdList writeIdList, Compa
     }
     StringBuilder extraDebugInfo = new StringBuilder("[").append(obsoleteDirs.stream()
         .map(Path::getName).collect(Collectors.joining(",")));
-    return remove(location, ci, obsoleteDirs, true, fs, extraDebugInfo);
+    boolean success = remove(location, ci, obsoleteDirs, true, fs, extraDebugInfo);
+    if (dir.getObsolete().size() > 0) {
+      updateDeltaFilesMetrics(ci.dbname, ci.tableName, ci.partName, obsoleteDirs);
```
Comment (klcopp) [inline ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/Cleaner.java:426]: I regret suggesting that we include aborted directories in the obsolete count. 
1. There are other metrics about aborted directories.
2. previouslyActiveDeltas - (obsolete + aborted) != currentlyActiveDeltas, so the active delta count would be off.

My bad :/

Comment (lcspinter) [reply ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/Cleaner.java]: Reverted.
